File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.py
from collections import defaultdict
import math
def solution(fees, records):
    r = defaultdict(list)
    
    for record in records:
        re = record.split(" ")
        h, m = re[0].split(":")
        time = int(h) * 60 + int(m)
        r[int(re[1])].append(time)
    
    answer = []
    for key in sorted(r.keys()):
        times = r[key]
        if len(times) % 2 == 1:
            times += [23 * 60 + 59]
        
        time = 0
        for i in range(0, len(times), 2):
            time += times[i+1] - times[i]
            
        time -= fees[0]
        result = fees[1]
        
        if time > 0:
            result += math.ceil(time/fees[2]) * fees[3]
        
        answer.append(result)
    return answer
        
        
        
        
        
        
        

File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.java
import java.util.*;
import java.util.stream.*;

class Solution {
    public int[] solution(int[] fees, String[] records) {
        Map<String, Integer> times = new HashMap<>();
        Map<String, Integer> park = new HashMap<>();
        
        for (String record : records) {
            String[] s = record.split(" ");
            int minute = changeTime(s[0]);
            String num = s[1];
            
            if(s[2].equals("IN")) {
                park.put(num, minute);
            } else {
                int inTime = park.remove(num);
                times.merge(num, minute - inTime, (o, v) -> o + v);
            }
        }
        int maxTime = changeTime("23:59");
        
        for (Map.Entry<String,Integer> kv : park.entrySet()) {
            times.merge(kv.getKey(), maxTime - kv.getValue(), (o, v) -> o + v);
        }

        return times.entrySet().stream()
            .sorted(Map.Entry.comparingByKey())
            .mapToInt(entry -> fees[1] + (int) Math.ceil(Math.max(0, times.get(entry.getKey()) - fees[0]) / (double)fees[2]) * fees[3])
            .toArray();
    }
    
    public int changeTime(String time) {
        String[] t = time.split(":");
        return Integer.parseInt(t[0]) * 60 + Integer.parseInt(t[1]);
    }
}